lib/chef/event_dispatch/dispatcher.rb

```diff
@@ -28,6 +28,9 @@ def register(subscriber)
       # Define a method that will be forwarded to all
       def self.def_forwarding_method(method_name)
         define_method(method_name) do |*args|
+          if method_name == :deprecation && args.size == 1
```

Comment:
deprecation() is the only thing that has an argument with a default relating to the caller, and we need to trigger that default early in the process.

When we have a better error matcher thingy on the other end--something that will figure out what recipe was invoked, for example--we can have that logic scan the stack and skip the dispatcher.